### ISSUE TITLE
Inconsistent clear in `apply ... in ... as ...` with intro patterns

### ISSUE THREAD

**Issue by Villetaneuse** [coq/coq#17210]
#### Description of the problem

Using `apply ... in I as J` does not clear `I`, which is certainly a good thing (why would you use `as` if not to keep `I`?)
When one performs an intro pattern in the `as` part, however, the `in` term is cleared:
```coq
Lemma and_comm_simpl (A B : Prop) : A /\ B -> B /\ A.
Proof. easy. Qed.

Lemma plop1 (A B : Prop) : A /\ B -> A /\ B.
Proof.
  intros H.
  apply and_comm_simpl in H as H'. (* ok, keeps H and H' *)
  easy.
Qed.

Lemma plop2 (A B : Prop) : A /\ B -> A /\ B.
Proof.
  intros H.
  apply and_comm_simpl in H as [hb ha]. (* not ok, does not keep H *)
  easy.
Qed.

Lemma plop3 (A B : Prop) : A /\ B -> A /\ B.
Proof.
  intros H.
  apply and_comm_simpl in H as H'%proj1. (* not ok, does not keep H *)
Admitted.

Lemma plop4 (n m : nat) : n + m = 0 -> m + n = 0.
Proof.
  intros H.
  apply eq_sym in H as ->. (* not ok, does not keep H *)
Admitted.
```

Is this behavior wanted? The dev version of the refman states :
If specified, as [simple_intropattern] is applied to the conclusion of the type of [one_term].
In this case, the selected hypothesis is left unchanged if its name is not reused.

(The stable version of the refman does not really mention the fact that the hypothesis could be left unchanged).

I think the hypothesis should remain, whether one uses an intro pattern or not, and the behaviour should be consistent.
#### Coq Version

This happens in upstream (8.18+alpha) and in 8.16.1.

**Comment by herbelin** [comment]
I agree that the keep behavior of `as H` is ad hoc, thus surprising, but this was the original intent. In particular, without `as` the hypothesis is not kept either.

What is missing is a better syntax than `as H'` to keep the hypothesis which is compatible with the presence of an intro pattern is used. Maybe `apply foo in kept H`/`apply foo in kept H as pat`? It is a bit long but clear.

**Comment by SkySkimmer** [comment]
With destruct we can do `destruct (H)` which syntactically turns the hypothesis into a term and terms are not cleared.
Maybe `apply foo in (term) as H` would make sense to have?

**Comment by Villetaneuse** [comment]
Imho, the most simple approach for the user is to ensure that the hypothesis is not cleared when the `as` form is used. The user can always `clear` if that's what she wants. Maybe it breaks too many things?

**Comment by herbelin** [comment]
> The user can always clear if that's what she wants.

Conversely, the user can always do `pose proof H as H'` to keep a copy of `H` before using `apply in`, so, the question to me is first what default we want (keep or clear), and, second, whether we can provide a lighter "on-the-fly" way to change the default.

For instance, with `apply` you can say `apply >H` to keep (though this is now deprecated and I don't know what are the alternative plans), and with `destruct` you can say `destruct (H)` to keep, as said by @SkySkimmer.

**Comment by Villetaneuse** [comment]
The `pose proof` beforehand is too long imho and feels too much like fighting with the system. If `destruct` already has this feature, maybe use it with apply for consistency (+ it's concise enough). But things become complicated :
- `apply foo in H as H'` : what happens to H?
- `apply foo in (H) as H'` : why the parentheses, I already said `as`.
- `apply foo in H as proper_intro_pattern` : clear H ?
- `apply foo in (H) as proper_intro_patter` : ok
Items 1, 2 and 3 are still problematic.
By the way, I love intro patterns and `apply` is great once you get to know it (hence the frustration and the issue), so please take it as an attempt of constructive criticism.

**Comment by Villetaneuse** [comment]
Also, clearing and keeping are not symmetric... keeping should be easier that cleaning since it doesn't lose information.

**Comment by herbelin** [comment]
>     * `apply foo in H as H'` : what happens to H?
>     * `apply foo in (H) as H'` : why the parentheses, I already said `as`.

If the syntax `in (H)` is introduced, then I guess `in H as H'` would clear `H` (as when an `as` is given).
 
>     * `apply foo in H as proper_intro_pattern` : clear H ?
>     * `apply foo in (H) as proper_intro_patter` : ok
>       Items 1, 2 and 3 are still problematic.

Assuming `H as H'` is changed to erase `H` by default, would 3 still be problematic in your eyes?

>       By the way, I love intro patterns and `apply` is great once you get to know it

Thanks.

**Comment by herbelin** [comment]
> keeping should be easier that cleaning since it doesn't lose information.

We can also argue that what should be easiest to do is the most common behavior.

**Comment by Villetaneuse** [comment]
I still think that the simplest behaviour is `as` does not clear (simplest to document, simplest to teach).
If it's not chosen, maybe the next best choice is:
- let everything as it is (no need to break anything)
- add parentheses to keep hypothesis in any situation
- document
